### PR TITLE
Add clustering function - RCFSummarize

### PR DESCRIPTION
**Pull request by model-collapse**
### Description
Adding RCFSummarize clustering algo into ml-commons
Issue Link here: https://github.com/opensearch-project/ml-commons/issues/356 

### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [* ] New functionality includes testing.
  - [* ] All tests pass
- [* ] New functionality has been documented.
  - [* ] New functionality has javadoc added
- [* ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/MathUtil.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.opensearch.ml.engine.utils;
+
+import java.util.function.BiFunction;
+public class MathUtil {
+    public static Integer iamconst = 1;
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/MathUtil.java:5]: Where this const being used? Add some comments ?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            parameters.getMaxK(), 
+            parameters.getInitialK(), 
+            parameters.getPhase1Reassign(), 
+            distance, 
+            new Random().nextLong(), 
+            parameters.getParallel());
+        
+        Iterable<float[]> centroidsLst = Arrays.asList(summary.summaryPoints);
+        List<Integer> predictions = new ArrayList<>();
+        Arrays.stream(featureNamesValues.v2()).forEach(e->predictions.add(MathUtil.findNearest(e, centroidsLst, distance)));
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java:153]: Just curious, I guess RCF has already calculated nearest centroid for all data points. Why doesn't RCF return the centroids for all data points directly (at least provide some option to let user decide whether output or not)? If so, we can avoid recalculating nearest centroids here.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    @Override
+    public MLOutput trainAndPredict(DataFrame dataFrame) {
+        Tuple<String[], float[][]> featureNamesValues = TribuoUtil.transformDataFrameFloat(dataFrame);
+        SampleSummary summary = Summarizer.summarize(featureNamesValues.v2(), 
+            parameters.getMaxK(), 
+            parameters.getInitialK(), 
+            parameters.getPhase1Reassign(), 
+            distance, 
+            new Random().nextLong(),
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java:148]: Minor: how about reusing some singleton `Random` object? Any reason to create a new `Random` object for each request?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/SerializableSummary.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.ml.engine.utils;
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/SerializableSummary.java:5]: Seems this class only used by RCF summarizer. How about move it to `.../algorithms/clustering` package?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/MathUtil.java (hunk 1 above)
Comment (model-collapse) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/MathUtil.java:5]: Will delete, copied from where else.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java (hunk 3 above)
Comment (model-collapse) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java:148]: Agree

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/SerializableSummary.java (hunk 4 above)
Comment (model-collapse) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/SerializableSummary.java:5]: Ok, will

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java (hunk 2 above)
Comment (model-collapse) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java:153]: Let me dig into the code, I am not sure if the library supports exporting these things.

Comment (model-collapse) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java:153]: Opps, they did export these things, but it is a good idea to follow.

Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java:154]: Thanks, I  created an issue to track this https://github.com/aws/random-cut-forest-by-aws/issues/338

**Review by ylwu-amzn — APPROVED** [review]

**Comment by ylwu-amzn** [comment]
DCO check failed, you can add `--signoff` when commit change

**Review by Zhangxunmt — COMMENTED** [review]
add "-s" after your "git commit" so it will auto adds the Signed-off-by: xxx .

File: common/src/main/java/org/opensearch/ml/common/input/parameter/clustering/RCFSummarizeParams.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    // The distance function
+    private DistanceType distanceType;
+    // Whether to also use Reassign in Phase1
+    private Boolean phase1Reassign;
+    // Whether to train in parallel
+    private Boolean parallel;
+    // TODO: expose seed?
+
+    @Builder(toBuilder = true)
+<<<<<<< HEAD
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/input/parameter/clustering/RCFSummarizeParams.java:53]: Need to resolve merge conflict.

File: common/src/main/java/org/opensearch/ml/common/input/parameter/clustering/RCFSummarizeParams.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                    initialK = parser.intValue(false);
+                    break;
+                case PHASE1_REASSIGN_FIELD:
+                    phase1Reassign = parser.booleanValue();
+                    break;
+                case PARALLEL__FIELD:
+                    parallel = parser.booleanValue();
+                    break;
+                case DISTANCE_TYPE_FIELD:
+                    distanceType = DistanceType.from(parser.text());
```
[the remark this answers is not in the document]
Comment (model-collapse) [reply common/src/main/java/org/opensearch/ml/common/input/parameter/clustering/RCFSummarizeParams.java:108]: Why upper case?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/KMeans.java [hunk 7]
```diff
@@ -31,16 +31,25 @@
 import java.util.Map;
 import java.util.Optional;
 
+/**
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/KMeans.java:34]: Remove this comment, the issue fixed.

**Review by ylwu-amzn — APPROVED** [review]
LGTM. Thanks for adding this algorithm!

**Review by jngz-es — APPROVED** [review]
LGTM, thanks.

**Pull request merged**